### PR TITLE
Feature/dev 459 ct key filter

### PR DESCRIPTION
**Pull request by Yannick1712**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations
  - Login/logout
  - Registration
  - Update user data
  - Create/delete payment routes
- [ ] **Wait for changes on payment page**

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

File: src/config/config.ts [hunk 1]
```diff
@@ -23,7 +23,15 @@ export class Configuration {
...
-  apiKeyVersionCT = '0'; // single digit hex number
+  apiKeyVersionCT = '1'; // single digit hex number
+
+  HistoryFilter = {
+    buy: 0,
+    sell: 1,
+    staking: 2,
+    ref: 3,
+    lm: 4,
+  };
```
Comment (davidleomay) [inline src/config/config.ts:34]: I would place this as an enum next to api-key.service.

File: src/shared/services/api-key.service.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import { Util } from '../util';
+
+@Injectable()
+export class ApiKeyService {
+  constructor() {}
+
+  public createKey(address: string, queryFilter: HistoryFilter): string {
+    let hash = Util.createHash(Util.createHash(address + new Date().toISOString(), 'sha256'), 'md5').toUpperCase();
+
+    return hash.substring(0, hash.length - 5) + this.getFilterHash(queryFilter) + Config.apiKeyVersionCT;
```
Comment (davidleomay) [inline src/shared/services/api-key.service.ts:15]: We should try to store this magic numbers (in this case 5) into variables.

File: src/shared/services/api-key.service.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+              : filter[key] != null
+              ? filter[key]
+              : null),
+      );
+
+      return filter;
+    }
+  }
+
+  private getFilterHash(queryFilter: HistoryFilter): string {
```
Comment (davidleomay) [inline src/shared/services/api-key.service.ts:55]: It's actually not a hash, more a code 😉

File: src/shared/services/api-key.service.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    let filterHashSubString = null;
+    Object.entries(queryFilter).map(([key]) => (filterHashSubString += Math.pow(2, Number(Config.HistoryFilter[key]))));
+    filterHashSubString = '0000' + (filterHashSubString ? filterHashSubString.toString(16) : '');
+    filterHashSubString = filterHashSubString.substring(filterHashSubString.length - 4, filterHashSubString.length);
+    return filterHashSubString;
```
Comment (davidleomay) [inline src/shared/services/api-key.service.ts:60]:
```ts
const filterCode = Util.sum(Object.keys(queryFilter).map((key) => Math.pow(2, Config.HistoryFilter[key])));
return filterCode.toString(16).padStart(4, '0');
```

Comment (davidleomay) [reply src/shared/services/api-key.service.ts:60]: As we use a hex code here, 2 or 3 digits should probably be enough.

File: src/shared/services/api-key.service.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+          (filter[key] =
+            (apiKeyFilter & Math.pow(2, value)) > 0
+              ? ((apiKeyFilter & Math.pow(2, value)) > 0).toString()
+              : filter[key] != null
+              ? filter[key]
+              : null),
+      );
+
+      return filter;
+    }
```
Comment (davidleomay) [inline src/shared/services/api-key.service.ts:52]: I would remove the filter parameter and just return the HistoryQuery resulting from the parsing. In the calling function you can use object destructuring to merge the queries (`const a = {...b, ...c};`)

File: src/shared/services/api-key.service.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+      return filter;
+    } else if (apiKeyVersion == '1') {
+      Object.entries(Config.HistoryFilter).map(
+        ([key, value]) =>
+          (filter[key] =
+            (apiKeyFilter & Math.pow(2, value)) > 0
+              ? ((apiKeyFilter & Math.pow(2, value)) > 0).toString()
+              : filter[key] != null
+              ? filter[key]
+              : null),
```
Comment (davidleomay) [inline src/shared/services/api-key.service.ts:48]:
```ts
Object.entries(Config.HistoryFilter).filter(([key, value) => apiKeyFilter & Math.pow(2, value) > 0).reduce((prev, [key, value]) => {...prev, [key]: 'true', {});
```

File: src/user/models/user/user.service.ts [hunk 7]
```diff
@@ -38,6 +40,7 @@ export class UserService {
     private readonly walletService: WalletService,
     private readonly settingService: SettingService,
     private readonly dfiTaxService: DfiTaxService,
+    private readonly apiKeyCT: ApiKeyService,
```
Comment (davidleomay) [inline src/user/models/user/user.service.ts:43]: Rename to apiKeyService?

File: src/shared/services/api-key.service.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    if (!user.apiKeyCT) throw new BadRequestException('API key is null');
+    return Util.createHash(user.apiKeyCT + user.created, 'sha256').toUpperCase();
+  }
+
+  public getSign(user: User, timestamp: string): string {
+    const secret = this.getSecret(user);
+    return Util.createHash(secret + timestamp, 'sha256').toUpperCase();
+  }
+
+  public checkKeySign(user: User, sign: string, timestamp: string): boolean {
```
Comment (davidleomay) [inline src/shared/services/api-key.service.ts:28]: Rename to `isValid` or similar? Would make the meaning of the result clear.

File: src/shared/services/api-key.service.ts [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    const userSign = this.getSign(user, timestamp);
+
+    return sign.toUpperCase() == userSign || Util.daysDiff(new Date(timestamp), new Date()) <= 1;
```
Comment (davidleomay) [inline src/shared/services/api-key.service.ts:31]: Are you sure that this is right? 😉

File: src/shared/services/api-key.service.ts (hunk 2 above)
Comment (Yannick1712) [reply src/shared/services/api-key.service.ts:15]: You mean as a variable in config?

**Review by davidleomay — APPROVED** [review]

File: src/shared/services/api-key.service.ts [hunk 10]
```diff
@@ -9,7 +9,7 @@ export class ApiKeyService {
   private versionLength = 1;
   private filterLength = 3;
   private get suffixLength(): number {
-    return this.versionLength + this.filterLength;
+    return this.versionLength;
```
Comment (davidleomay) [inline src/shared/services/api-key.service.ts:12]: suffixLength can be removed completely and version can be back to '0'.

File: src/shared/services/api-key.service.ts [hunk 11]
```diff
@@ -38,25 +38,26 @@ export class ApiKeyService {
...
   // --- KEY HANDLING --- //
-  public create(address: string, filter: HistoryFilter): string {
+  public createKey(address: string): string {
     const hash = Util.createHash(Util.createHash(address + new Date().toISOString(), 'sha256'), 'md5').toUpperCase();
-    return hash.substring(0, hash.length - this.suffixLength) + this.filterToCode(filter) + Config.apiKeyVersionCT;
+    return hash.substring(0, hash.length - this.suffixLength) + Config.apiKeyVersionCT;
   }
 
-  public getFilter(apiKey: string): HistoryFilter {
+  public getFilter(apiKey: string, filterCode: string): HistoryFilter {
```
Comment (davidleomay) [inline src/shared/services/api-key.service.ts:46]: I think the filter is not depending on the (version of the) apiKey, is it?

**Review by davidleomay — APPROVED** [review]

**Pull request merged**